BUG: Patch Checked Add Method

gfyoung opened the pull request:
Title is self-explanatory.  Follow-up to #14237.

Merged in PR #14453

gfyoung committed:
BUG: Catch overflow in both directions for checked add

1) Add checks to ensure that add overflow does not
   occur both in the positive or negative directions.
2) Add benchmarks to ensure that operations involving
   this checked add function are significantly impacted.

jorisvandenbossche commented:
Although this is probably more correct, I don't think the performance degradation is allowable:

```
# master

In [1]: arr = np.arange(1000000)
   ...: td = pd.to_timedelta(arr)
   ...: ts = pd.Timestamp('2000')
   ...: 

In [2]: %timeit td + ts
100 loops, best of 3: 2.51 ms per loop
```

```
# PR

In [2]: %timeit td + ts
10 loops, best of 3: 41.1 ms per loop
```

The problem is in `np.iinfo(np.int64).min - b > arr` when `b` is positive (and probably in the other check of `b` is negative)

gfyoung commented:
@jorisvandenbossche : I draw a hard line on correctness.  This is impeding my ability to patch the other TDI overflow issues because there are tests that fail because of this bug.  Perhaps it would instead be worthwhile to break the check up into two separate statements, though not sure that would help too much given that it's Python.

jreback commented:
try this

```
In [5]: arr = np.arange(1000000)
   ...: td = pd.to_timedelta(arr)
   ...: ts = pd.Timestamp('2000')
   ...: a = td.view('i8')
   ...: b = ts.value
   ...: 

In [7]: nmax = np.iinfo(np.int64).max 

In [8]: nmin = np.iinfo(np.int64).min

n [11]: ((b > 0) & (np.iinfo(np.int64).max - b < arr)).any() or((b < 0) & (np.iinfo(np.int64).min - b > arr)).any()
Out[11]: False

In [12]: %timeit ((b > 0) & (np.iinfo(np.int64).max - b < arr)).any() or((b < 0) & (np.iinfo(np.int64).min - b > arr)).any()
10 loops, best of 3: 57.7 ms per loop

In [14]: %timeit pd.eval('((b > 0) & (nmax - b < arr)) | ((b < 0) & (nmin - b > arr))').any()
100 loops, best of 3: 6.24 ms per loop

In [15]: pd.eval('((b > 0) & (nmax - b < arr)) | ((b < 0) & (nmin - b > arr))').any()
Out[15]: False

```

jreback commented:
numba can do even better (but we don't have this in the quiver atm)

```
In [17]: def f(a,b,arr,nmax,nmin):
    ...:     return np.any((b > 0) & (nmax - b < arr)) | ((b < 0) & (nmin - b > arr))

In [18]: f = numba.jit(f,nopython=True)

In [19]: %timeit f(a,b,arr,nmax,nmin).any()
The slowest run took 59.62 times longer than the fastest. This could mean that an intermediate result is being cached.
100 loops, best of 3: 2.28 ms per loop

In [20]: f(a,b,arr,nmax,nmin).any()
Out[20]: False
```

jorisvandenbossche commented:
Something like this is quite convoluted, but is a bit faster as the eval approach:

```
In [62]: %%timeit
    ...: b2 = np.broadcast_to(b, arr.shape)
    ...: mask1 = b2 > 0
    ...: mask2 = b2 < 0
    ...: if not mask1.any():
    ...:     to_raise = (np.iinfo(np.int64).min - b2 > arr).any()
    ...: elif not mask2.any():
    ...:     to_raise = (np.iinfo(np.int64).max - b2 < arr).any()
    ...: else:
    ...:     to_raise = (np.iinfo(np.int64).max - b2[mask1] < arr[mask1]).any() or (np.iinfo(np.int64).min - b2[mask1] > arr[mask1]).any()
    ...: 
100 loops, best of 3: 4.2 ms per loop
```

gfyoung commented:
@jorisvandenbossche : Are you proposing I use this more convoluted method then?  Thoughts?

gfyoung commented:
@jorisvandenbossche : Some feedback on my question <a href="https://github.com/pydata/pandas/pull/14324#issuecomment-251452745">above</a> would be greatly appreciated.

jreback commented:
@gfyoung yes pls use one of the methods proposed above

gfyoung commented:
@jorisvandenbossche : Your solution is unfortunately not backwards compatible with older versions of `numpy` that don't have `np.broadcast_to`.

jreback commented:
so put a conditional on the numpy version (or use the eval soln)

gfyoung commented:
@jreback : I thought we don't do those kind of checks (or try to avoid them).  In any case, I'm trying with another method as I mentioned above.

jreback commented:
i don't see a reason not to do a check in the numpy version

or use the eval soln (which is fine but will fall back to numpy if numexpr is not installed)

gfyoung commented:
@jreback : Fair enough.  Also, I have no objections to your eval solution.  I'm trying it right now.

jorisvandenbossche commented:
The `eval` solution will be very slow if you don't have numexpr installed, so I would prefer not to use it.

gfyoung commented:
@jorisvandenbossche : Unfortunately, your solution is not backwards compatible.  And if you don't like @jreback suggestion, then what do you suggest?  Compared to my original change, the `eval` solution would at least be an improvement.

jreback commented:
just check the numpy version

i am not sure why you think this is an issue

gfyoung commented:
@jreback : And then what, use the slower solution that I used before if the version is not compatible?

jreback commented:
yep

gfyoung commented:
@jreback : How about my second question?

jorisvandenbossche commented:
I am sure that there is a way to achieve the same as `broadcast_to` does for older numpy versions. Eg it uses `np.nditer` under the hood. Or we can simply check for scalar vs aryray: scalar -> fill array of correct shape with this, array -> should have the same shape as the other array

gfyoung commented:
@jorisvandenbossche : Your fill idea could work actually.  Let me see what I can do there.

gfyoung commented:
Well at least the `pd.eval` method works, but I'll do the `numpy` version check now.

shoyer commented:
broadcast_to is only a few lines of pure Python code. It would be quite straightforward to put it in a company module -- that's what  I did in xarray: https://github.com/pydata/xarray/blob/master/xarray/core/npcompat.py

gfyoung commented:
@shoyer : I suppose, but why boilerplate that if a simpler solution can be used?

gfyoung commented:
New timing information:
~~~python
In [1]: import numpy as np

In [2]: import pandas as pd

In [3]: arr = np.arange(1000000)

In [4]: td = pd.to_timedelta(arr)

In [5]: ts = pd.Timestamp('2000')

# Master
In [6]: %timeit td + ts
100 loops, best of 3: 3.51 ms per loop

# PR
In [6]: %timeit td + ts
100 loops, best of 3: 4.06 ms per loop
~~~

gfyoung commented:
@jreback , @jorisvandenbossche : Travis is passing, and the performance is significantly better than it was initially.  Ready to merge if there are no other concerns.

jreback commented on pandas/core/nanops.py:
```diff
@@ -829,9 +830,28 @@ def _checked_add_with_arr(arr, b):
-        raise OverflowError("Python int too large to "
-                            "convert to C long")
+    if np.__version__ >= LooseVersion('1.10.0'):
+        b2 = np.broadcast_to(b, arr.shape)
```
i would put a comment back to this issue and s short expl of why this is being done (perf)

it's not obvious to someone looking at this code why (or even ehat) is happenening

gfyoung replied:
Sure thing.  Done.

jreback commented on pandas/core/nanops.py:
```diff
@@ -829,9 +830,42 @@ def _checked_add_with_arr(arr, b):
-                            "convert to C long")
+    # For performance reasons, we broadcast 'b' to the new array 'b2'
+    # so that it has the same size as 'arr'.
+    if np.__version__ >= LooseVersion('1.10.0'):
```
use the pd _version_undernp11 that are already defined (look up the spelling of the variable)

gfyoung replied:
What do you mean?  I can't find anything matching that name you provided at the moment.  Do you know where such a variable used?

jorisvandenbossche replied:
https://github.com/pydata/pandas/blob/master/pandas/compat/numpy/__init__.py#L14

gfyoung replied:
Ah, that's the name!  Will add in place of the explicit `LooseVersion`.

jorisvandenbossche commented on pandas/core/nanops.py:
```diff
@@ -829,9 +830,42 @@ def _checked_add_with_arr(arr, b):
+        to_raise = (np.iinfo(np.int64).max - b2 < arr).any()
+    else:
+        to_raise = ((np.iinfo(np.int64).max - b2[mask1] < arr[mask1]).any() or
+                    (np.iinfo(np.int64).min - b2[mask1] > arr[mask1]).any())
```
I suppose one of these two lines should use mask2 instead of mask1 (probably an error in my example code)

jorisvandenbossche commented:
Can you add benchmarks that have all possible cases with possible performance hits? (positive overflow, negative overflow, combination -> the three branches of the if/elif/else in the check)

gfyoung commented:
@jorisvandenbossche : Which benchmark suite should I place it under?

jorisvandenbossche commented:
@gfyoung I would add specific benchmarks for the `_checked_add_with_arr` with the three overflow cases to algorithms.py or binary_ops.py, and then add a normal case (no overflow) to timedelta.py (eg the example used in the discussion here)

gfyoung commented:
@jorisvandenbossche : For each overflow case, I presume it's just `try-except`?

jorisvandenbossche commented:
@gfyoung Sorry, on second thought: not with actual overflow cases (because that indeed errors, not needed to put in a benchmark I think), but with positive, negative, and mixed values in `b` (which are the cases for which the if/elif/else branches are introduced)

gfyoung commented:
@jorisvandenbossche : Ah, okay.  So cases that will cause each branch to get checked but not err.  Got it.

gfyoung commented:
@jreback , @jorisvandenbossche : Added the benchmarks, and Travis and Appveyor are passing.  Ready to merge if there are no other concerns.

jorisvandenbossche commented on pandas/core/nanops.py:
```diff
@@ -829,9 +830,42 @@ def _checked_add_with_arr(arr, b):
+            b2 = np.empty(arr.shape)
+            b2.fill(b)
+        else:
+            b2 = b.copy()
```
I don't think this copy is needed?

jorisvandenbossche replied:
Is the alias a problem? It only exists for the scope of this function? (`broadcast_to` also returns a view, no copy)

Now it just takes an unnecessary copy of an array just to sum it, which has further performance implications (a quick test with the above example codes, it gives me almost 20% of time of the full method)

gfyoung replied:
Hmmm...okay.  I'll give it a try and see what happens.

jorisvandenbossche commented on pandas/core/nanops.py:
```diff
@@ -829,9 +830,42 @@ def _checked_add_with_arr(arr, b):
+    # gh-14324: The implementation below takes advantage of NumPy
+    # broadcasting to check all of these elements at once. Although
+    # it is somewhat convoluted, it has much better performance than
+    # say iterating through all elements in both 'arr' and 'b2'.
```
Maybe instead of this comment, I would clarify that the checks have to be performed separately for negative/positive overflow for performance reasons

(the fact that iterating over an array in pure python is not performant should be rather obvious to developers I think)

gfyoung replied:
@jorisvandenbossche : Perhaps, but I see no reason why we should make that assumption regarding array iteration.  I prefer to leave as is but will include a comment about separating out the checks.

gfyoung replied:
@jorisvandenbossche : I modified the documentation as requested, and the build machines seem happy at the moment.

gfyoung replied:
When multiple reviewers request the change, fair enough.  Done.

Commit:
BUG: fix json with lines=True for quoted special characters

closes #14391
closes #14390

chris-b1 committed:
BLD/CI: cython cache pxd files (#14363)

jreback committed:
DOC: formatting in basics.rst

jreback committed:
BUG: Bug in localizing an ambiguous timezone when a boolean is passed

closes #14402
closes #14405

adamchainz committed:
Convert readthedocs links for their .org -> .io migration for hosted projects (#14406)

As per [their blog post of the 27th April](https://blog.readthedocs.com/securing-subdomains/) ‘Securing subdomains’:

> Starting today, Read the Docs will start hosting projects from subdomains on the domain readthedocs.io, instead of on readthedocs.org. This change addresses some security concerns around site cookies while hosting user generated data on the same domain as our dashboard.

Test Plan: Manually visited all the links I’ve modified.

jorisvandenbossche commented:
This looks good to me. 
@gfyoung final question: can you run the timedelta benchmark (`test_add_td_ts`) and compare this to 0.18 as well?

jreback commented:
lgtm

gfyoung commented:
@jorisvandenbossche : Using the same setup as <a href="https://github.com/pandas-dev/pandas/pull/14324#issuecomment-252539410">above</a>, 100 loops, best of 3: 3.08 ms per loop

jorisvandenbossche commented:
And compared to 0.18? (on my machine I get 1.15 ms, so probably something of 3x slowdown)

gfyoung commented:
@jorisvandenbossche : That timing information is for 0.18.0

jorisvandenbossche commented:
@gfyoung That is a bit strange, because then you have 4.06 for this PR and 3.08 for 0.18. I get 5.11 vs 1.11, so a much larger difference.

Anyway, it's a rather large performance degredation, but we decided to go down this path.

jreback committed:
DOC: pydata/pandas -> pandas-dev/pandas (#14409)

jreback committed:
PERF: improved perf in .to_json when lines=True

closes #14408

jonmmease committed:
Bug: Grouping by index and column fails on DataFrame with single index (GH14327) (#14428)

Existing logic under "if level is not None:" assumed that index was a MultiIndex.
Now we check and also handle the case where an Index is passed in with a None grouper.
This resolves GH 14327

chris-b1 committed:
PERF: Period factorization (#14419)

gfyoung committed:
CLN: Removed coerce param in pd.to_timedelta and pd.to_datetime

Deprecated back in `v0.17.0` <a href="https://github.com/pydata/pandas
/commit/987b7e7e586b8df1d127406c69e0a9094a1a5322">here</a>.  Seems
appropriate to carry though now.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13602 from gfyoung/remove-coerce-to-timedelta and squashes the following commits:

091af83 [gfyoung] CLN: Removed coerce param in pd.to_timedelta

brandonmburroughs committed:
BUG: Concat with axis rows

closes #14369

Author: Brandon M. Burroughs <brandonmburroughs@gmail.com>

Closes #14416 from brandonmburroughs/concat_with_axis_rows and squashes the following commits:

9aa260d [Brandon M. Burroughs] Changing axis handling to depend on object passed
49442be [Brandon M. Burroughs] Using dataframe _get_axis_number instance method
64702fb [Brandon M. Burroughs] Updating documentation for concat
fdd5260 [Brandon M. Burroughs] Removing duplicate expected dfs
3f08b07 [Brandon M. Burroughs] Adding concat tests for axis 0, 1, and 'index'
cf3f998 [Brandon M. Burroughs] Adding ValueError test for concat Series axis 'columns'
a6694b9 [Brandon M. Burroughs] Updating documentation
584ebd2 [Brandon M. Burroughs] BUG: Allow concat to take string axis names

jreback commented:
@gfyoung you may need to close this and open a new PR. It seems that people are unable to rebase and push correctly to the original pandas repo. (further appevey was disconnected) so might be a good test.

gfyoung commented:
@jreback : Ah, I can see that now.  Will close and repost.

jorisvandenbossche commented:
@jreback BTW, I contacted github about it, and it should be solved now (was an error on their side). So the closing/open new PR should no longer be needed for all older PRs.

jreback commented:
@jorisvandenbossche good to know! thanks!

gfyoung commented:
A little to late now but oh well. :smile:  New PR in #14453.
